### PR TITLE
Add Provenance field in TaskRun&PipelineRun status

### PR DESCRIPTION
**Pull request by chuangw6**
# Changes

/kind feature

resolve https://github.com/tektoncd/pipeline/issues/5550

**Change 1:** Add a `Provenance` field in TaskRun&PipelineRun status. This field
currently only contains a subfield named `ConfigSource`, but can be extended later to
have more provenance-related fields.

**Change 2:** Prior, https://github.com/tektoncd/pipeline/pull/5551 introduced the `ConfigSource` to api/resolution alpha & beta 
package. In this PR, we moved the `ConfigSource` to api/pipeline alpha & beta package so that 
the `Provenance` struct can reuse the `ConfigSource` type (Otherwise, we cannot import it from the api/resolution alpha because of the import cycle).

**Why:** See the motivation and discussions in https://github.com/tektoncd/pipeline/issues/5550.
The tl;dr is that it helps pass provenance-related data in a more structured way
ConfigSource is one example.

Signed-off-by: Chuang Wang <chuangw@google.com>

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Add Provenance field in TaskRun&PipelineRun status that wraps all the information we might need from pipeline side.

It only contains ConfigSource at the moment, but it can be extended to have more subfields in future.
```

### CODE REVIEW

**Comment by chuangw6** [comment]
Hey @abayer ,
FYI: I am moving the forked `ConfigSource` type from api/resolution to api/pipeline so that we can reuse that type in the `Provenance` struct. (Otherwise there will be import cycle issue.) Thanks!

**Comment by abayer** [comment]
Thanks for letting me know, @chuangw6 - if this gets in before #5515 (which is likely), I'll update my PR accordingly!

File: pkg/apis/pipeline/v1/provenance.go
Comment (XinruZhang) [inline pkg/apis/pipeline/v1/provenance.go:18]: Nit: Identifies

Comment (chuangw6) [reply pkg/apis/pipeline/v1/provenance.go]: Changed. Thanks @XinruZhang !

File: pkg/apis/pipeline/v1beta1/provenance.go
Comment (Yongxuanzhang) [inline pkg/apis/pipeline/v1beta1/provenance.go:25]:
```suggestion
	// Field Definition: https://github.com/in-toto/attestation/blob/main/spec/field_types.md#ResourceURI
```

Comment (Yongxuanzhang) [reply pkg/apis/pipeline/v1beta1/provenance.go:25]: I thought this is an example at first glance

**Comment by abayer** [comment]
Sorry, @chuangw6 - #5515 got in before this. =)

**Comment by chuangw6** [comment]
> Sorry, @chuangw6 - #5515 got in before this. =)

Great! I synced this PR with latest main. All good!!  Thanks for letting me know @abayer !

File: pkg/apis/pipeline/v1beta1/provenance.go
Comment (chuangw6) [reply pkg/apis/pipeline/v1beta1/provenance.go]: Sounds good! Changed. Thanks @Yongxuanzhang ~

**Comment by chuangw6** [comment]
> /LGTM though I don't actually have the permission, thanks Chuang!

Thank you @XinruZhang for taking time to review this PR! Appreciate it :D

**Review by dibyom — APPROVED** [review]
/approve

This LGTM! But given that we are adding a new field to status I wanted to see if any of the other @tektoncd/core-maintainers had any feedback on this

File: pkg/apis/pipeline/v1/provenance.go
Comment (jagathprakash) [inline pkg/apis/pipeline/v1/provenance.go:19]: Rename the variable Source as ConfigSource, so that it matches the name in SLSA specification https://slsa.dev/provenance/v0.2.

Comment (jagathprakash) [inline pkg/apis/pipeline/v1/provenance.go:19]: nit: Name the variable as ConfigSource, so that it matches the name in the SLSA specification. https://slsa.dev/provenance/v0.2

**Comment by chuangw6** [comment]
> Update https://github.com/tektoncd/pipeline/blob/main/docs/pipeline-api.md#tekton.dev/v1.PipelineRunStatus as well?

Good catch!! Thank you @jagathprakash!! Updated!

File: pkg/apis/pipeline/v1/provenance.go
Comment (chuangw6) [reply pkg/apis/pipeline/v1/provenance.go]: > nit: Name the variable as ConfigSource, so that it matches the name in the SLSA specification. https://slsa.dev/provenance/v0.2

Good idea! I like the idea. Renamed!  Thank you @jagathprakash :D

**Comment by chuangw6** [comment]
RE: This Monday's API WG Meeting

Hi @afrittoli @jerop @dibyom @lbernick ,
Thanks for sharing your thoughts in the meeting around putting this behind a feature flag. This sounds good to me! I am happy to add it behind a dedicated feature flag i.e. `enable-provenance-status`with only "true" and "false" options. 

But this shouldn't block this PR for now since this PR doesn't not write a value to the provenance field yet. I'll make sure it's behind the feature flag when I write a value to the field in next [PR](https://github.com/tektoncd/pipeline/pull/5397). 

Thank you!!

**Comment by chuangw6** [comment]
/assign @wlynch 

PTAL! Thank you!

File: pkg/apis/pipeline/v1/pipelinerun_types.go
Comment (wlynch) [inline pkg/apis/pipeline/v1/pipelinerun_types.go:431]: I'd improve the docs here - this is tautological (i.e. "a foo is a foo") and references what looks like another type that does exist (assume users don't know about SLSA).

File: pkg/apis/pipeline/v1/provenance.go
Comment (wlynch) [inline pkg/apis/pipeline/v1/provenance.go:22]:
```suggestion
// ConfigSource identifies the source where a resource came from.
// This can include Git repositories, Task Bundles, file checksums, or other information
// that allows users to identify where the resource came from and what version was used.
```

File: pkg/apis/pipeline/v1/pipelinerun_types.go
Comment (wlynch) [reply pkg/apis/pipeline/v1/pipelinerun_types.go:431]: (also in the other locations where this comment is used)

**Review by wlynch — CHANGES_REQUESTED** [review]
Up to you if you want to do it this for this PR: You did a good job identifying 2 separate changes that are included in this PR - if you find that things can be split up neatly into 2 separate issues like this I'd always recommend splitting into multiple PRs accordingly. 🙏 

It's much easier to review because you don't have to parse which line goes with which change.

File: pkg/apis/pipeline/v1/pipelinerun_types.go
Comment (chuangw6) [reply pkg/apis/pipeline/v1/pipelinerun_types.go]: Great point. I just updated the document for this field and the type as well. Please take a look. Happy to update it further for better doc!
Thanks @wlynch

File: pkg/apis/pipeline/v1/provenance.go
Comment (chuangw6) [reply pkg/apis/pipeline/v1/provenance.go]: Great doc for the field. Updated. Thanks @wlynch

**Comment by chuangw6** [comment]
> Up to you if you want to do it this for this PR: You did a good job identifying 2 separate changes that are included in this PR - if you find that things can be split up neatly into 2 separate issues like this I'd always recommend splitting into multiple PRs accordingly. 🙏
> 
> It's much easier to review because you don't have to parse which line goes with which change.

@wlynch Great suggestion. Thank you for the review and the feedback!! Really appreciate it!

For this PR particular, I'd prefer to keep it as is if it's okay for now b/c using configsource type in provenance struct is the main reason why we move it to from resolution api pkg to pipeline api pkg. IMHO, doing the move in another PR (i.e. move it to a provenance file without actually using it in that pkg) might be a bit confusing :D 

But you raised a good point like this might make it hard for reviewers to parse the mapping. In future, I'll improve on this i.e. open 2 PRs and link to each other in the description. 

Thank you very much!!!

**Review by wlynch — APPROVED** [review]
/lgtm

**Comment by chuangw6** [comment]
resolve #5550

**Pull request merged**
